### ISSUE TITLE
Spec issue: GridSample

### ISSUE THREAD

**Issue by broune** [onnx/onnx#4033]
This is a collection of issues encountered with GridSample in the ONNX spec.

It's not clear to me if this is the proper method/format/venue to report these - let me know if I should be doing something else.

For reference: [current ONNX spec for GridSample](https://github.com/onnx/onnx/blob/main/docs/Operators.md#GridSample) and [PyTorch GridSample](https://pytorch.org/docs/stable/generated/torch.nn.functional.grid_sample.html)

I also have [a draft spec change proposal](https://github.com/broune/onnx/blob/main/docs/Operators.md#GridSample), which simply makes some assumptions about how the issues below will be resolved and so would likely need some updates later. Regardless of those decisions, this draft is far more detailed in explaining the behavior of GridSample than anything else I've seen, and may be useful for anyone seeing this who has not spent a lot of time looking at GridSample recently. (I also can't figure out how to display LaTeX equations in the spec, as you will notice if you read that link, is LaTeX possible / allowed in the spec?)

## Errors
1) The description of how the reflection padding_mode works for points close to the border is true for align_corners=0 but not when align_corners=1 (which moves the axis of mirroring, so it’s different). Both the Microsoft ONNX implementation and PyTorch do something different for align_corners=1 and doing so is also consistent with the description in the spec of how this works for points far from the border.

2) The spec allows the grid input to contain complex numbers without defining what that means. I don’t know what it would mean to access an array at a complex number (with or without interpolation) and I think allowing this is an error.

3) The X input is allowed to be a complex number (which is fine), but the output type specification does not include a complex number, which seems like an error, since complex number input would yield complex number output (assuming the obvious interpretation of how complex number input should be handled).

## Missing details
4) The spec does not specify how normalized coordinates are denormalized.

5) The spec does not specify that the minor coordinate comes before the major coordinate, though it can be seen from the first example (which matches PyTorch). This is surprising and opposite from the order in gather_nd, so would be good to specify explicitly.

6) It’s not specified what the element type of the output should be. If the X input is floating point, it makes sense that the output would use the same element type, but this is not specified anywhere. Also, in that case, what precision should internal calculations use? This is a bigger question for integer inputs - it’s really not clear what the output type should then be. Presumably some floating point type.
 
7) The spec doesn't say anything about what happens in the padding mode "nearest" when there is a tie where a grid point is equally close to two or more integer points. This shows up as a denormalized coordinate ending in ".5" and the question is whether to round it up or down. There is an example provided in the spec using "nearest" padding and that example has several ties in it. If we leave this unspecified, so that any implementation is allowed, it would be good for the spec to say that and for the provided example in the spec to not contain this case (it currently does) so that the correct answer is unambiguous. Note that the Resize op does specify how rounding should be done, providing several options, we could specify the same.
 
8) This op is unusual in that inherently-imprecise floating point calculations are used as indices into arrays. If the grid points are so large that one ulp exceeds 1.0, then the results can be arbitrarily imprecise and there is a problem for implementations casting to integer points. The spec ought to address this, I’d prefer by saying that such large grid points can yield arbitrary results, freeing implementations from having to do arbitrary-precision calculations to fit the spec.
 
## Minor points
9) The grid operand is allowed to be integer. This will hit the grid only at integer points (corner points), which it is not clear that anyone would want to do. If we continue to allow this, it might be good for the spec to acknowledge that it’s supported but probably would indicate a mistake to use it.

10) The X operand is allowed to be a complex number. This works out easily mathematically, but it might be good for the spec to acknowledge that the simple interpretation of treating the real and imaginary components as independent dimensions is the intended interpretation. (Unless we didn’t intend to support complex numbers here at all?)

## Further points for discussion
11) ONNX only allows 4D points for GridSample but PyTorch also allows 5D inputs. Conceivably 3D inputs with a time dimension might get into 6D. Should we allow 5D dimensions? How about 6D dimensions?

12) Beyond rounding mode, Resize also allows setting the alpha parameter for bicubic interpolation. Should we allow alpha (and rounding mode, but there’s a separate point about that above) to be set for bicubic interpolation on GridSample as well?

**Comment by broune** [comment]
Three more missing details:

13. The spec does not specify what happens if the input width or height are zero (one of them or both). That's straightforward to define for the zeros padding mode (everything is zero), though it's harder for nearest and border. Either way, the spec should say whether it is allowed and, if it is, how it is supported.

14. The spec specifies what happens when align_corners is 0 or 1. It does not say what happens for other values. Presumably other values are not allowed and, if so, it would be good for the spec to specify that.

15. The spec states that the align_corners attribute is an "int", but it doesn't say what kind of int - how many bits and whether it is signed (this is not unique to GridSample). Perhaps that means that the attribute's type can be any integer type, or maybe it means that the type is platform-dependent and must match the notion of "int" in some system compiler? I didn't find the meaning of "int" addressed in some other part of the spec, but maybe I'm just missing it.

**Comment by garymm** [comment]
@hwangdeyu can you please take a look? If the points raised seem valid to you, please work with @broune to open a PR with the proposed changes.

**Comment by hwangdeyu** [comment]
> @hwangdeyu Deyu Huang FTE can you please take a look? If the points raised seem valid to you, please work with @broune to open a PR with the proposed changes.

That's a good proposal. Will look at it in detail.

**Comment by hwangdeyu** [comment]
Thanks @broune for the extremely detailed collection. Could you create a PR for your spec change proposal? Then we can make a review with other contributors who were attended the kernal implementation.
If you are not convenient, please just tell me, then I will create an one.

**Comment by broune** [comment]
I'm happy to make a spec proposal, especially since I almost already did so. Just to avoid some unnecessary editing work, maybe we can have a preliminary (non-binding) idea of:

A) whether we want to support 5D and 6D? I think 5D ought to happen. 6D I don't know.

B) Whether we want to be able to set alpha.

C) Whether we want to be able to set rounding mode.

D) What do I do about math in the spec proposal? It's not rendered in github preview. Do we just not allow math formulas in the spec? Or is there a different way that I should do it? (this seems a big limitation for a math-heavy topic like ML)

E) I don't know how to resolve 15. This seems to be standard operating procedure across the spec, so I'm not sure what the fix is.

**Comment by hwangdeyu** [comment]
A) So far we only have the 4D kernel implementation, so we shouldn't add 5D spec description now.
B\C) I think it's good to add those firstly.
D) I'm not sure why the math formulas are not display in the spec markdown doc. You can also provide it as your draft proposal and I will ask the docs owner review and their suggestion of how to write formulas better.
E) Attributes are used for values that are constants in the graph. For each operator attribute, we must set a type for it in ONNX standard. Even it's an `int64` type, we only use `align_corners` for `0` or `1`. Cause we don't have the `boolean` type for attributes.

**Comment by broune** [comment]
Here is the pull request for this spec change, where I just resolved the open issues here as I saw fit (except I did not add 5D/6D GridSample, but I think we should have that): https://github.com/onnx/onnx/pull/4089 For the intent of creating discussion on something concrete as suggested further up in this thread.

I also noticed an additional spec issue:

16. The type constraint T1 includes a tensor of strings, which I just removed in my proposal, as it seems like a mistake.

**Comment by thiagocrepaldi** [comment]
Would https://github.com/pytorch/pytorch/issues/69674 leverage this? We need Pytorch' s `::grid_sampler` to export a detectron2 model

**Comment by hwangdeyu** [comment]
> Would [pytorch/pytorch#69674](https://github.com/pytorch/pytorch/issues/69674) leverage this? We need Pytorch' s `::grid_sampler` to export a detectron2 model

Hi @thiagocrepaldi , [`GridSimple`](https://github.com/onnx/onnx/blob/main/docs/Operators.md#GridSample) is already a operator in ONNX sepc now. This issue proposes the spec changes of `GridSample` op in ONNX.
Once PyTorch-ONNX enables opset 16, `grid_sampler` will be exported directly.
So for, we can use it as a contrib op in ort. Here is the usage. https://github.com/microsoft/onnxruntime/blob/6fb29f5b9a50f95e18c81a2052c00bac2f23e3dd/onnxruntime/python/tools/pytorch_export_contrib_ops.py#L36

**Comment by thiagocrepaldi** [comment]
@gramalingam Have you seen this?

**Comment by thiagocrepaldi** [comment]
@gramalingam @jackwish gentle ping

**Comment by zhenhuaw-me** [comment]
@thiagocrepaldi Thank you for reminding me. Unfortunately, I am not familiar with GridSample. Could check if the author of the GridSample spec to help to address this issue?

**Comment by gramalingam** [comment]
Thanks for the reminder Thiago! Need to find someone knowledgeable to review the PR, will take a look at it.

**Comment by gramalingam** [comment]
Re. points 2, 3, 9, and 10: these are valid points. The typing of the op doesn't look right. I suggest 
1. Changing the grid-type and output-type to be the same, which can be any floating-point type.
2. Dropping complex-types from the input-type.

The reasoning is that:
* Complex types are not supported much in ONNX as it is.
* If the input type is allowed to be complex, then the output-type does have to be extended to include complex type.
* Complex type does not seem appropriate for grid-type, in any case.
* The existing type-inference, which pushes the input type to output type seems incorrect. We can modify it to push the grid-type to output-type. (Alternatively, if the output-type is independent, then we need to add an attribute to specify the desired output-type.

**Comment by liqunfu** [comment]
> A) So far we only have the 4D kernel implementation, so we shouldn't add 5D spec description now.
> 
> B\C) I think it's good to add those firstly.
> 
> D) I'm not sure why the math formulas are not display in the spec markdown doc. You can also provide it as your draft proposal and I will ask the docs owner review and their suggestion of how to write formulas better.
> 
> E) Attributes are used for values that are constants in the graph. For each operator attribute, we must set a type for it in ONNX standard. Even it's an `int64` type, we only use `align_corners` for `0` or `1`. Cause we don't have the `boolean` type for attributes.

We do need to add 5D support. PyTorch has 5D case: https://pytorch.org/docs/stable/generated/torch.nn.functional.grid_sample.html?highlight=grid_sample

**Comment by yang-Michael** [comment]
When will the 5D input supported version be released？

**Comment by mattmagic149** [comment]
Hi,

are there any updates on the 5D support?

Cheers,
Matt

**Comment by sithu31296** [comment]
Any update to 5D input?

**Comment by justinchuby** [comment]
Marked for discussion. Anyone interested please feel free to chime in

**Comment by xyh19960622** [comment]
Any update to 5D input?

**Comment by thiagocrepaldi** [comment]
@liqunfu FYI

**Comment by henryruhs** [comment]
We need this feature, please implement it.

**Comment by andife** [comment]
@xadupre Is a short-term fix to be expected, or postponement to 1.19?
